Add unit tests for IArrayCreationExpression and IArrayInitializer

mavasani opened the pull request:
Fixes #17596

Issues identified while adding unit tests: https://github.com/dotnet/roslyn/issues/18150

mavasani commented:
Tagging @dotnet/analyzer-ioperation @AlekseyTs @cston for review.

genlu commented:
Do we want to test erroneous code as well?

mavasani commented:
@genlu I have added a unit test for invalid array creation in this PR: https://github.com/dotnet/roslyn/pull/17938/files#diff-b955da5b79c539293b33f4d4a23510deR336

I will add some more invalid creation tests.

mavasani commented:
@genlu Added unit tests for error cases.

mavasani commented:
@heejaechang @333fred @jinujoseph - can you please take a look?

333fred commented on src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_ArrayCreationAndInitializer.cs:
```diff
@@ -0,0 +1,415 @@
+{
+    public void F(char dimension)
+    {
+        var a = /*<bind>*/new M[dimension]/*</bind>*/;
```
Can you also include some error tests with dimensions that are not implicitly convertible, but are used as implicitly convertible anyway?

333fred replied:
Also tests for the result being a method call, with no conversion, correct/incorrect implicit conversion, correct/incorrect explicit conversion?

333fred commented on src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_ArrayCreationAndInitializer.cs:
```diff
@@ -0,0 +1,415 @@
+    public void F()
+    {
+        var a = """";
+        var b = /*<bind>*/new[] { ""hello"", a, null }/*</bind>*/;
```
Could you also include tests for arrays created with explicit sizes and content initializers, correct and incorrect?

333fred commented on src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_ArrayCreationAndInitializer.vb:
```diff
@@ -0,0 +1,435 @@
+﻿' Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
```
Do we need to add `ReDim` tests here?

AlekseyTs replied:
I would test ReDim statement separately, I am not sure if we even have representation for it in the current IOperation model.

333fred commented on src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_ArrayCreationAndInitializer.vb:
```diff
@@ -0,0 +1,435 @@
+
+            Dim expectedOperationTree = <![CDATA[
+IArrayCreationExpression (Dimension sizes: 1, Element Type: M) (OperationKind.ArrayCreationExpression, Type: M())
+  IBinaryOperatorExpression (BinaryOperationKind.IntegerAdd) (OperationKind.BinaryOperatorExpression, Type: System.Int32, Constant: 2)
```
Why in the heck is this an `IBinaryExpression`?!

AlekseyTs replied:
That is how the language works, if you specified 1, the size of the allocated array will be two (one is added). I believe this is for compatibility with VB6, which had one-based arrays. So, it was common to have VB6 code to allocate an array of size N and then index into the array using N as index.

333fred commented on src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_ArrayCreationAndInitializer.vb:
```diff
@@ -0,0 +1,435 @@
+
+            Dim expectedOperationTree = <![CDATA[
+IArrayCreationExpression (Dimension sizes: 1, Element Type: M) (OperationKind.ArrayCreationExpression, Type: M())
+  IBinaryOperatorExpression (BinaryOperationKind.IntegerAdd) (OperationKind.BinaryOperatorExpression, Type: System.Int32)
```
Again with the `IBinaryExpression`

333fred commented on src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_ArrayCreationAndInitializer.vb:
```diff
@@ -0,0 +1,435 @@
+
+            Dim expectedOperationTree = <![CDATA[
+IArrayCreationExpression (Dimension sizes: 3, Element Type: System.Byte) (OperationKind.ArrayCreationExpression, Type: System.Byte(,,))
+  IBinaryOperatorExpression (BinaryOperationKind.IntegerAdd) (OperationKind.BinaryOperatorExpression, Type: System.Int32, Constant: 1)
```
Lots of `IBinaryOperationExpression`s.

333fred commented on src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_ArrayCreationAndInitializer.vb:
```diff
@@ -0,0 +1,435 @@
+
+            Dim expectedOperationTree = <![CDATA[
+IArrayCreationExpression (Dimension sizes: 1, Element Type: System.Int32(,)) (OperationKind.ArrayCreationExpression, Type: System.Int32()(,))
+  IBinaryOperatorExpression (BinaryOperationKind.IntegerAdd) (OperationKind.BinaryOperatorExpression, Type: System.Int32, Constant: 1)
```
Again

333fred commented on src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_ArrayCreationAndInitializer.vb:
```diff
@@ -0,0 +1,435 @@
+
+            Dim expectedOperationTree = <![CDATA[
+IArrayCreationExpression (Dimension sizes: 2, Element Type: System.String) (OperationKind.ArrayCreationExpression, Type: System.String(,), IsInvalid)
+  IBinaryOperatorExpression (BinaryOperationKind.IntegerAdd) (OperationKind.BinaryOperatorExpression, Type: System.Int32, Constant: 2)
```
And again. In this case, the resulting IOp tree here is very different from the C# version. Is this another result of the highest/lowest difference?

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_ArrayCreationAndInitializer.cs:
```diff
@@ -0,0 +1,415 @@
+}
+";
+            string expectedOperationTree = @"
+IArrayCreationExpression (Dimension sizes: 1, Element Type: System.String) (OperationKind.ArrayCreationExpression, Type: System.String[])
```
>System.String [](start = 60, length = 13)

A question for design team, is there a value in duplicating the element type in a special property, can't we get it from the type for the node?

mavasani replied:
This PR went dormant and I forgot to file an issue for this. Opened https://github.com/dotnet/roslyn/issues/22265

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_ArrayCreationAndInitializer.cs:
```diff
@@ -0,0 +1,415 @@
+            VerifyOperationTreeForTest<ArrayCreationExpressionSyntax>(source, expectedOperationTree);
+        }
+    }
+}
```
Consider also adding a test for array creation of the form ```int[] x = { 1, 2 };```.

AlekseyTs replied:
Consider also testing 
```
new[] ;
new[2] ;
```
with and without target type.

AlekseyTs commented on src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_ArrayCreationAndInitializer.vb:
```diff
@@ -0,0 +1,435 @@
+  IBinaryOperatorExpression (BinaryOperationKind.IntegerAdd) (OperationKind.BinaryOperatorExpression, Type: System.Int32, Constant: 2)
+    Left: ILocalReferenceExpression: dimension (OperationKind.LocalReferenceExpression, Type: System.Int32, Constant: 1)
+    Right: ILiteralExpression (OperationKind.LiteralExpression, Type: System.Int32, Constant: 1)
+  IArrayInitializer (OperationKind.ArrayInitializer)
```
>IArrayInitializer (OperationKind.ArrayInitializer) [](start = 2, length = 50)

It feels like there should be no initializer for case like this.

AlekseyTs commented on src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_ArrayCreationAndInitializer.vb:
```diff
@@ -0,0 +1,435 @@
+  IBinaryOperatorExpression (BinaryOperationKind.IntegerAdd) (OperationKind.BinaryOperatorExpression, Type: System.Int32)
+    Left: IParameterReferenceExpression: dimension (OperationKind.ParameterReferenceExpression, Type: System.Int32)
+    Right: ILiteralExpression (OperationKind.LiteralExpression, Type: System.Int32, Constant: 1)
+  IArrayInitializer (OperationKind.ArrayInitializer)
```
>IArrayInitializer (OperationKind.ArrayInitializer) [](start = 2, length = 50)

It feels like there should be no initializer for a case like this. I see this as a case where presence of ```{}``` is just a syntax requirement, in fact it is an error to have it non-empty.

AlekseyTs commented on src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_ArrayCreationAndInitializer.vb:
```diff
@@ -0,0 +1,435 @@
+            Dim source = <![CDATA[
+Class C
+    Public Sub F()
+        Dim a = {{1, 2, 3}, {4, 5, 6}}'BIND:"{{1, 2, 3}, {4, 5, 6}}"
```
>{{1, 2, 3}, {4, 5, 6}} [](start = 16, length = 22)

Do we have tests for erroneous array literals (mismatched element types, mismatched dimensions, etc.)?

AlekseyTs commented on src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_ArrayCreationAndInitializer.vb:
```diff
@@ -0,0 +1,435 @@
+
+            VerifyOperationTreeForTest(Of ArrayCreationExpressionSyntax)(source, expectedOperationTree)
+        End Sub
+    End Class
```
>End Class [](start = 4, length = 9)

Consider testing cases like these:
```
        x = New Integer(2) {1, 2}
        x = New Integer(x(0)) {1, 2}

        Dim y = New Integer(x(0), x(1)) {}
        y = New Integer(1, 1) {}
        y = New Integer(1, 1) {{}}
        y = New Integer(1, 1) {{}, {}}
        y = New Integer(1, 1) {{1, 2}}
        y = New Integer(1, 1) {{1, 2}, {}}
        y = New Integer(1, 1) {{1, 2}, {1, 2}}
```

AlekseyTs replied:
Are we testing these scenarios?

---
In reply to: [108760949](https://github.com/dotnet/roslyn/pull/18154#discussion_r108760949) [](ancestors = 108760949)

mavasani replied:
Done - sorry I had missed this comment.

AlekseyTs commented:
How should the following array creation be exposed?
```
Dim x(2) As Integer
```

Perhaps it should be covered variable declarations. CC @333fred

mavasani commented:
@333fred @AlekseyTs I have addressed your feedback and the PR is ready for review.

One additional API change has been done: I have removed `IArrayCreationExpression.ElementType` as per https://github.com/dotnet/roslyn/issues/22265

mavasani commented:
@dotnet-bot retest ubuntu_14_debug_prtest please

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_ArrayCreationAndInitializer.cs:
```diff
@@ -0,0 +1,853 @@
+}
+";
+            string expectedOperationTree = @"
+IArrayCreationExpression (OperationKind.ArrayCreationExpression, Type: System.Int32[], IsInvalid) (Syntax: 'new[2]/*</bind>*/')
```
>System.Int32[] [](start = 71, length = 14)

It feels strange that we decided the type the array is int[]. Please open an issue to investigate.

mavasani replied:
https://github.com/dotnet/roslyn/issues/22315

AlekseyTs commented:
We should have a test for this declaration ```Dim x(2) As Integer```

AlekseyTs commented:
Done with review pass (iteration 8)

mavasani commented:
@AlekseyTs - added requested unit tests.

mavasani commented:
@cston @jcouv @dotnet/roslyn-compiler Can I have one more review please?
